### PR TITLE
[FSDP2] Added `mesh` arg, `FSDPState`, move to device

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118584
* #118298
* #118223
* #118198
* #118188
* #118136
* #118118
* #118017
* #118004
* #117975
* #117973
* #117955
* #117950
* #117881
* #117877
* #117867
* __->__ #117814
* #118525

Squashed to include https://github.com/pytorch/pytorch/pull/117861, https://github.com/pytorch/pytorch/pull/117852

---

This PR adds `_get_managed_modules()` to determine which modules a `fully_shard(module)` call manages. The rule is defined as:
> `fully_shard(module)` manages all modules in `module.modules()` except those already managed by a nested `fully_shard()` or a nested non-composable API (e.g. `replicate()` or TorchRec).

Practically, this can be implemented as a graph search from `module` that does not proceed into any module with `fully_shard` or a non-composable API applied. Because the non-composable APIs follow the same rule, this rule is correct inductively.

---

This PR adds `_get_managed_states(managed_modules)` to return the managed parameters and buffers given the managed modules.
- Without an extra mechanism to ignore specific parameters or buffers, the rule currently is simply to get the directly managed state (i.e. parameters/buffers) from each managed module while de-duplicating shared ones.
- However, we prefer this translation from managed modules to managed states to accommodate ignoring specific states in the future (which has appeared in various open-source use cases).

---

This PR adds the `mesh` argument to `fully_shard` and some helper data structures specific to FSDP/HSDP that pre-compute useful info like rank/world size for each mesh dim.
- The `mesh` defines the FSDP/HSDP algorithm. 1D mesh means FSDP, and 2D mesh means HSDP, where we assume sharding on the last dimension.
    - We can revisit the HSDP sharding-dim assumption if needed in the future.
- The default (if `mesh is None`) is that `fully_shard` calls `init_device_mesh` following the global process group.
- The helper data structures are the various `*MeshInfo`s. I included up to the `HSDPMeshInfo` even though it will not be immediately used to show the spirit of it. We want to tag both the shard and replicate dims.
- The `mesh_info` variable in `fully_shard` is not used for now. It will be passed downstream in the future.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: torch/distributed/_composable/fsdp/_fsdp_init.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+from typing import Union
+
+import torch
+
+
+def _normalize_device(device: Union[torch.device, int, str]) -> torch.device:
```
Comment (wconstab) [inline torch/distributed/_composable/fsdp/_fsdp_init.py:6]: this looks fine, but, why is it an fsdp-specific util, do we not have anything like this in torch already, or maybe we should push it that direction if not?

**Comment by wconstab** [comment]
are the [FSDP2] PRs related to landing per-param stuff? looks like this applys the contract to fully_shard too, is that just nice to have or also important for some use case?

File: torch/distributed/_composable/fsdp/_fsdp_init.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_init.py:6]: I agree that this is not FSDP-specific. I have not seen a function like this available in Python, but I may have missed it.

I am happy to move this elsewhere if we have an idea of where that should be.

Comment (wconstab) [reply torch/distributed/_composable/fsdp/_fsdp_init.py:6]: there is actually a similar one in torch/distributed/pipeline/sync/worker.py and a non-similar one in test/functorch/test_eager_transforms.py

i would propose to put something in the vicinity of actual torch.Device code to help with this.  cc @ezyang @zou3519 

we can do it in another PR

Comment (Skylion007) [inline torch/distributed/_composable/fsdp/_fsdp_init.py:6]: We have a type Alias for that here: https://github.com/pytorch/pytorch/blob/af831415a876079a2ff63d72ae99d03e4c11aebd/torch/_prims_common/__init__.py#L55

File: torch/distributed/_composable/fsdp/fully_shard.py [hunk 2]
```diff
@@ -1,21 +1,30 @@
...
+from ._fsdp_init import _normalize_device
+from ._fsdp_state import FSDPState
 
-@contract()
+
+@contract(state_cls=FSDPState)
 def fully_shard(
     module: nn.Module,
+    *,
+    device: Union[torch.device, int, str] = "cuda",
```
Comment (Skylion007) [inline torch/distributed/_composable/fsdp/fully_shard.py:17]: Likewise, we have a TypeAlias: https://github.com/pytorch/pytorch/blob/af831415a876079a2ff63d72ae99d03e4c11aebd/torch/_prims_common/__init__.py#L55

File: torch/distributed/_composable/fsdp/_fsdp_init.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_init.py:6]: I changed it to use this type alias. 

The rebase conflicts when changing the imports (e.g. touching `from typing import ...`) are rough 😢 .

Comment (ezyang) [reply torch/distributed/_composable/fsdp/_fsdp_init.py]: sure. bikeshed a name lol

File: torch/distributed/_composable/fsdp/fully_shard.py [hunk 3]
```diff
@@ -3,21 +3,30 @@
...
 ):
     if isinstance(module, (nn.ModuleList, nn.ModuleDict)):
         raise ValueError(
             f"fully_shard does not support containers that do not implement forward: {module}"
         )
+    device = _normalize_device(device)
     state = fully_shard.state(module)
     _insert_module_state(module, state)
+    state._module = module
+    state._device = device
```
Comment (weifengpy) [inline torch/distributed/_composable/fsdp/fully_shard.py:29]: nit: state init seems be scatterred vertically. is it possible to have ``fully_shard.state(*args, **kwargs)`` where we put can pack ``.module=`` ``_device`` logic into ``fully_shard.state(*args, **kwargs)`` ?

File: torch/distributed/_composable/fsdp/_fsdp_init.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+from torch._prims_common import DeviceLikeType
+
+
+def _normalize_device(device: DeviceLikeType) -> torch.device:
+    if isinstance(device, torch.device):
+        if device == torch.device("cuda"):
+            return torch.device("cuda", torch.cuda.current_device())
+        return device
+    elif isinstance(device, int):
+        return torch.device("cuda", device)
```
Comment (weifengpy) [inline torch/distributed/_composable/fsdp/_fsdp_init.py:12]: it looks generic enough to be a pytorch-level utils except this line. We seems to assume ``int/1`` referring to ``cuda:1``?

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_init.py:12]: I think that assumption is already made in `torch`:
```
>>> import torch
torch>>> torch.device(1)
device(type='cuda', index=1)
```

File: torch/distributed/_composable/fsdp/fully_shard.py (hunk 3 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/fully_shard.py:29]: Good point. I think the only constraint is that the state object is constructed in the `contract` decorator, so we cannot use the `FSDPState.__init__()` directly.

Talking out loud: Even if we have a separate `FSDPState.init()` function (distinct from `__init__()`), we will eventually have to have 1 line per assignment like `self._module = module` and `self._device = device`, etc. I am not sure if there is a way around that. What do you think?

File: torch/distributed/_composable/fsdp/_fsdp_init.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import torch
+
+from torch._prims_common import DeviceLikeType
+
+
+def _normalize_device(device: DeviceLikeType) -> torch.device:
+    if isinstance(device, torch.device):
+        if device == torch.device("cuda"):
+            return torch.device("cuda", torch.cuda.current_device())
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/_fsdp_init.py:9]: looks like this still relies on cuda or `torch.cuda` to be present, to support device agnostic, should we add the `device_handle` logic to getattr from the `torch` namespace?

File: torch/distributed/_composable/fsdp/fully_shard.py (hunk 3 above)
Comment (wanchaol) [reply torch/distributed/_composable/fsdp/fully_shard.py:29]: I feel a separate `FSDPState.init()` would be cleaner, and we can put the `_insert_module_state` to the state class itself? this makes relatively clean separation between state init and fsdp core logic

File: test/distributed/_composable/fsdp/test_fully_shard_init.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+    @property
+    def world_size(self) -> int:
+        return 1
+
+    def test_normalize_device_cpu(self):
+        for device in ("cpu", torch.device("cpu")):
+            self.assertEqual(_normalize_device(device), torch.device("cpu"))
+
+    def test_normalize_device_meta(self):
```
Comment (wanchaol) [inline test/distributed/_composable/fsdp/test_fully_shard_init.py:24]: nit: `cpu` `meta` looks exactly same logic, maybe we can put them in the same test case?

**Review by wanchaol — APPROVED** [review]
LGTM!

File: torch/distributed/_composable/fsdp/_fsdp_init.py (hunk 5 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_init.py:9]: Yes! I wanted to avoid the `device_handle` for now to move faster. I can only test CUDA, so I wanted to write everything first assuming CUDA. It should not be too hard to abstract it to `device_handle` in the future.

File: torch/distributed/_composable/fsdp/fully_shard.py (hunk 3 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/fully_shard.py]: Sounds good. I added an `FSDPState.init()`.

File: torch/distributed/_composable/fsdp/fully_shard.py [hunk 7]
```diff
@@ -1,25 +1,60 @@
...
+    mesh: Optional[DeviceMesh] = None,
 ):
+    """
+    Args:
+        mesh (Optional[DeviceMesh]): This mesh defines the sharding and device.
+            If this is a 1D mesh, then this fully shards across the 1D mesh
+            (i.e. FSDP). If this is a 2D mesh, then this shards across the 0th
+            dimension and replicates across the 1st dimension (i.e. HSDP).
+            FSDP/HSDP uses the device given by the mesh's device type. For CUDA
+            or CUDA-like devices, FSDP expects uses the current device.
```
Comment (wconstab) [inline torch/distributed/_composable/fsdp/fully_shard.py:36]: "FSDP expects uses the current device."  remove expects, i guess?

**Review by wconstab — APPROVED** [review]

File: torch/distributed/_composable/fsdp/fully_shard.py (hunk 7 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/fully_shard.py:36]: Yep! Thanks. Let me remove the "expects" the next time I push.

File: torch/distributed/_composable/fsdp/_fsdp_common.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+            dist.ProcessGroup, self.mesh.get_group(self.replicate_mesh_dim)
+        )
+        self.replicate_mesh_rank: int = self.replicate_process_group.rank()
+
+
+@dataclass
+class HSDPMeshInfo(FSDPMeshInfo, DDPMeshInfo):
+    def __post_init__(self):
+        super(FSDPMeshInfo, self).__post_init__()
+        super(DDPMeshInfo, self).__post_init__()
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_common.py:53]: These two lines should instead be:
```
super(HSDPMeshInfo, self).__post_init__()
```
Thanks @clee2000  for pointing this out!
